src/DNSC/DNSUtil.hs

```diff
@@ -9,22 +10,64 @@ import Control.Monad (void)
 import Time.System (timeCurrent)
 import qualified Time.Types as Time
 import Network.Socket (Socket, SockAddr)
+#if MIN_VERSION_network(3,1,2)
+import qualified Network.Socket as Socket
+#endif
 import qualified Network.Socket.ByteString as Socket
 import Network.DNS (DNSMessage)
 import qualified Network.DNS as DNS
 
 
 ---
 
--- receive DNSMessage from client with address
-recvFrom :: Socket -> IO (DNSMessage, SockAddr)
-recvFrom sock = do
-  let bufsiz = 16384 -- maxUdpSize in dns package, internal/Network/DNS/Types/Internal.hs
-  (bs, peer) <- Socket.recvFrom sock bufsiz `E.catch` \e -> E.throwIO $ DNS.NetworkFailure e
-  Time.Elapsed (Time.Seconds now) <- timeCurrent
-  case DNS.decodeAt now bs of
-    Left  e   -> E.throwIO e
-    Right msg -> return (msg, peer)
-
-sendTo :: Socket -> DNSMessage -> SockAddr -> IO ()
-sendTo sock = (void .) . Socket.sendTo sock . DNS.encode
+#if MIN_VERSION_network(3,1,2)
+type Cmsg = Socket.Cmsg
+#else
+type Cmsg = ()
+#endif
+
+mkRecv :: Bool -> Socket -> IO (DNSMessage, (SockAddr, [Cmsg]))
```

Comment:
Why don't you use `(DNSMessage, SockAddr, [Cmsg])`?

Reply:
simply destruct and assign tuples at each step

https://github.com/khibino/dns-resolver/commit/7ca7a7b232dc539acd300b56b23c2485a272bbf3

return tuples that can be reused in the request and response queues

https://github.com/khibino/dns-resolver/commit/55cacbd91782e8c8bfff2426eb4ab07072af4710

Which is better?